src/SingleSignatureDirtyHack.sol

```diff
@@ -0,0 +1,272 @@
+            mstore(0x40, chainid())
+            mstore(0x60, address())
+            r := keccak256(0x00, 0x80)
+            mstore(0x40, ptr)
+            mstore(0x60, 0x00)
+        }
+    }
+
+    function _DOMAIN_SEPARATOR() internal view returns (bytes32) {
+        return block.chainid == _cachedChainId ? _cachedDomainSeparator : _computeDomainSeparator();
+    }
+
+    function DOMAIN_SEPARATOR() external view returns (bytes32) {
+        return _DOMAIN_SEPARATOR();
+    }
+
+    function _hashStruct(string calldata typeSuffix, TransferParams calldata transferParams)
+        internal
+        view
+        returns (bytes32 signingHash)
+    {
+        bytes32 domainSep = _DOMAIN_SEPARATOR();
+        address operator = _msgSender();
+        uint256 deadline = transferParams.deadline;
+        bytes32 structHash = transferParams.structHash;
+
+        assembly ("memory-safe") {
+            let ptr := mload(0x40)
+
+            mstore(ptr, 0x5472616e73666572416e64286164)
+            mstore(add(0x20, ptr), "dress operator,uint256 deadline,")
+            calldatacopy(add(0x40, ptr), typeSuffix.offset, typeSuffix.length)
+
+            mstore(0x00, keccak256(add(0x12, ptr), add(0x2e, typeSuffix.length)))
+
+            // We don't bother to clean dirty bits here. We assume that the presence of dirty bits
+            // will either cause a revert elsewhere in this context or that the (hash of the) bits
+            // are signed over, rejecting any dirtiness.
+            mstore(0x20, operator)
+            mstore(0x40, deadline)
+            mstore(0x60, structHash)
+
+            structHash := keccak256(0x00, 0x80)
+
+            mstore(0x00, 0x1901)
+            mstore(0x20, domainSep)
+            mstore(0x40, structHash)
+
+            signingHash := keccak256(0x1e, 0x42)
+
+            mstore(0x60, 0x00)
+            mstore(0x40, ptr)
+        }
+    }
+
+    /// @inheritdoc IERC5267
+    function eip712Domain()
+        external
+        view
+        override
+        returns (
+            bytes1 fields,
+            string memory name_,
+            string memory,
+            uint256 chainId,
+            address verifyingContract,
+            bytes32,
+            uint256[] memory
+        )
+    {
+        fields = bytes1(0x0d);
+        name_ = name;
+        chainId = block.chainid;
+        verifyingContract = address(this);
+    }
+
+    // While the check on the allowance *SHOULD* be sufficient to prevent replay, we duplicate the
+    // protocol-level nonce at the application level to guard against weird ERC20s that might not
+    // decrease `allowance`. ERC20 does not require that `transferFrom` decrease the existing
+    // allowance.
+    function _consumeNonce(address owner, uint256 incomingNonce) private {
+        uint256 currentNonce = nonces[owner];
+        nonces[owner] = incomingNonce;
+        if (incomingNonce <= currentNonce) {
+            revert NonceReplay(currentNonce, incomingNonce);
+        }
```

Comment:
```suggestion
        if (incomingNonce <= currentNonce) {
            revert NonceReplay(currentNonce, incomingNonce);
        }
        nonces[owner] = incomingNonce;
```

Reply:
Solc is dumb and is only able to optimize away (sometimes, when it feels like it) the duplicate `nonces[owner]` hash in the former case. Since the latter only optimizes the reverting path, I think it's preferable to have it the way that it is now. Unless you're thinking of something else?